astropy/io/ascii/sextractor.py

```diff
@@ -36,14 +36,18 @@ def get_cols(self, lines):
         # However, some may be missing and must be inferred from skipped column numbers
         columns = {}
         # E.g. '# 1 ID identification number' (without units) or '# 2 MAGERR magnitude of error [mag]'
-        re_name_def = re.compile(r"""^ \s* \# \s*            # possible whitespace around #
+        # Updated along with issue #4603, for more robust parsing of unit
+        re_name_def = re.compile(r"""^\s* \# \s*             # possible whitespace around #
                                  (?P<colnumber> [0-9]+)\s+   # number of the column in table
                                  (?P<colname> [-\w]+)        # name of the column
-                                 (?:\s+(?P<coldescr> \w [^\[]*\w))? # column description, match non-[
-                                 (?:\s+\[(?P<colunit>.+)\])?.*   # match units in brackets
+                                 (?:\s+(?P<coldescr> \w .+)  # column description, match any character until...
+                                 (?:(?<!(\]))$|(?=(?:(?<=\S)\s+\[.+\]))))?  # ...until [non-space][space][unit] or [not-right-bracket][end]
```

Comment:
Just curious why this change to the regex parsing is needed.  Can you show an example of what was failing before?  I'm not saying this is wrong, just trying to understand.